### ISSUE TITLE
WAUTH-2 support

### ISSUE THREAD

**Issue by tvandergeer** [hidespb/node-whatsapi#11]
Very exiting project! Do you have any plans on implementing support for WAUTH-2?

**Comment by thiagupct** [comment]
@matteocontrini 
Difficulty - Diffie–Hellman key exchange encyption. 

If we need this then this has to be built again from scratch. If we provide partial support what is now in WhatsAPI, all messages we receive will be encrypted. So, am not ok with that.

Looking for reasons,
1. Why we should move to WAUTH-2? 
2. What are the advantages?

Node API is working fine and WhatsApp is not going to stop their support for Nokia S40 in near future. It will take some more years for them to fully support WAUTH-2 [Diffie–Hellman encryption]. When some repo has cracked the end-end encryption or WAUTH-1 goes off let's migrate then.

**Comment by matteocontrini** [comment]
The only thing I miss from WAUTH-2 at the moment is the contacts sync, which is sometimes needed (and I don't know if it's fine to use WhatsAPI and node API at the same time).

For the rest, everything's good for me right now, although some fixes are needed and I'll push them as pull requests in the next days.

I have no experience with encryption, so I don't know how much I could help D: I'll search and see what I understand. Has yowsup managed to decrypt messages?

To sum up, there is no particular reason, except contacts synching, for switching to WAUTH-2. Let's stay on -1 as long as we can, I'd say :)

**Comment by matteocontrini** [comment]
@thiagupct just seen these updates on WhatsAPI and Yowsup. I think they managed to handle the encryption.
- https://github.com/mgp25/WhatsAPI-Official/wiki/WhatsApp-incoming-updates
- https://github.com/tgalal/yowsup/wiki/Yowsup-2:-End-to-End-encryption

Anyway it seems that, for now, it's an Android-only thing.

**Comment by thiagupct** [comment]
@matteocontrini We can start porting from that python lib.

It's an Android-only thing and group, media messages are not yet encrypted.

**Comment by matteocontrini** [comment]
@thiagupct it's not easy.. I have no experience with encryption. Do you?

**Comment by thiagupct** [comment]
Not at all. Successful third month into programming. 

Don't worry we are going to learn a lot:-)

**Comment by matteocontrini** [comment]
@thiagupct so I have more experience than you :) I've been programming for some years now, even if I'm only 17.
I've wrote you an email yesterday. Did you get it?

I think we should start from WAUTH-2 without thinking about the encryption which will come to WhatsAPI soon, but that's not easy neither D: I'll see what I understand

**Comment by thiagupct** [comment]
Yes, check your mail.

@tvandergeer , @matteocontrini  Let's get started.

**Comment by tvandergeer** [comment]
In the last days i've been working on WAUTH-2 and WA 1.5 protocol support for node-whatsapi. So far I've been able to successfully authenticate using the WAUTH-2 mechanism and exchange a couple of messages. I've been doing this by porting implementations from WhatsAPI (PHP) and Yowsup. 
Currently I' stuck on some issue with decoding and incoming message after I sent a presence update right after authentication, but it should take a few more hours to solve this. 
I've got the feeling that I'm really close to have this first step ready. After this we probably have to review all the messages, because some attributes moved. And some new messages were introduced.

**Comment by tvandergeer** [comment]
I've pushed my code [here](https://github.com/tvandergeer/node-whatsapi/tree/features/WAUTH-2).
Warning: it's not finished yet, but it should give you an idea of the progress so far. Feel free to comment.

**Comment by tvandergeer** [comment]
I've found the problem and now it's working. See [the branch in my fork](https://github.com/tvandergeer/node-whatsapi/tree/features/WAUTH-2).

**Comment by thiagupct** [comment]
@tvandergeer Great work and support from you.

Am checking the stability of WAUTH-2. Reconnecting a lot for me, don't know why. 

So protocol.js is the pending one to be updated. Correct me if am wrong.

**Comment by matteocontrini** [comment]
Every part should be updated since some files haven't changed for more than one year. Everything needs to be checked with WAUTH-2.

**Comment by thiagupct** [comment]
As mentioned am getting all the messages, presence status, composing/paused status, receipt, acknowledgements with WAUTH-2. But the nodes have changed a lot. @tvandergeer - SUPERB
```javascript
//Message counter
{"contents":{"tag":"ib","attributes":{"from":"s.whatsapp.net"},"children":[{"contents":{"tag":"offline","attributes":{"count":"4"},"children":[],"data":""}}],"data":""}}

//Message
{"contents":{"tag":"message","attributes":{"from":"919876543210@s.whatsapp.net","id":"","offline":"0","type":"text","t":"1420348376","notify":"WOLF"},"children":[{"contents":{"tag":"body","attributes":{},"children":[],"data":[73,110,105,109,97,108,101,32,100,105,115,116,117,114,98,32,112,97,110,110,97,116,104,97]}}],"data":""}}

//Presence 
{"contents":{"tag":"presence","attributes":{"from":"919876543210@s.whatsapp.net"},"children":[],"data":""}}

//Composing
{"contents":{"tag":"chatstate","attributes":{"from":"919876543210@s.whatsapp.net"},"children":[{"contents":{"tag":"composing","attributes":{},"children":[],"data":""}}],"data":""}}

//Paused
{"contents":{"tag":"chatstate","attributes":{"from":"919876543210@s.whatsapp.net"},"children":[{"contents":{"tag":"paused","attributes":{},"children":[],"data":""}}],"data":""}}

//Message receipt
{"contents":{"tag":"receipt","attributes":{"from":"919876543210@s.whatsapp.net","id":"message-1420369597-1","t":"1420369623"},"children":[],"data":""}}

//Acks - Don't know why this is for. But should be useful - Need research
{"contents":{"tag":"ack","attributes":{"from":"9876543210@s.whatsapp.net","id":"message-1420348387-1","class":"message","t":"1420348399"},"children":[],"data":""}}
```
@matteocontrini , @tvandergeer Processnode method should be updated such that we can provide WAUTH-2 support. We can fix the issues as and when it is found. We can do the performance improvements or is there anything that should be done first before providing WAUTH-2 support?

With WAUTH-2 all the incoming messages should be provided with receipts [Double tick, Blue tick]. Testing that method will provide soon.
@matteocontrini We are lagging with the registration process. Have to finish that very very soon.

**Comment by tvandergeer** [comment]
I think the reconnects are caused by some missing response. Maybe some keep-alive mechanism needs to be implemented. 
I know the nodes have changed a lot. That's why I paused my work on the group functionality and went for WAUTH-2 + WA1.5 protocol support first.  Don't know wether the changes to the messages/nodes are caused by WAUTH-2 or WA1.5, but that doesn't really matter now anymore.

**Comment by matteocontrini** [comment]
A ping/pong mechanism should be already there

**Comment by thiagupct** [comment]
ping/pong mechanism was deprecated in WAUTH-1. @tvandergeer make PR for WAUTH-2

**Comment by matteocontrini** [comment]
But let's keep the branch.

**Comment by tvandergeer** [comment]
Matteo and I suggest to put all work on WAUTH-2 and WA1.5 on a separate branch until all messages work. This way we will prevent a (partially) broken master branch until we got it all sorted out

**Comment by tvandergeer** [comment]
WAUTH-2 + WA1.5 support is now available on branch [WAUTH-2_and_WA1.5](https://github.com/hidespb/node-whatsapi/tree/WAUTH-2_and_WA1.5)

Closed issue.
